Add a guide to using the API for auto-discovery

ptgott opened the pull request:
See #19716

Adding resources is a popular use of Teleport's API, so I have added a guide to using the API for syncing resources to a service discovery API using an example of a local Docker setup.

ptgott committed:
Add a guide to using the API for auto-discovery

See #19716

Adding resources is a popular use of Teleport's API, so I have added a
guide to using the API for syncing resources to a service discovery API
using an example of a local Docker setup.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+- Use the Teleport gRPC API to look up resources registered with Teleport.
+- For any resources in your infrastructure that are not registered with
+  Teleport, use the Teleport API to spin up a new Teleport service or register
+  the resource with an exisitng Teleport service.
```
```suggestion
  the resource with an exisiting Teleport service.
```

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+  the resource with an exisitng Teleport service.
+- For any resources that are registered with Teleport but not in your
+  infrastructure, use the Teleport API to deregister the resource and, if
+  necessary, remove the Teleport service proxying the resource.
```
In the desktop registration [example](https://github.com/gravitational/teleport/tree/master/examples/desktop-registration) I've instead allowed items to expire automatically, so this step isn't necessary. I find it makes the example easier, but up to you.

ptgott replied:
I wanted to make the removal close to immediate for the demonstration of the app in the final section, but I'll add a note that you can change the TTL.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+
+- Fetching RabbitMQ management API endpoints registered with Teleport.
+- Fetching RabbitMQ containers.
+- If a RabbitMQ container does not conrrespond to a management API endpoint,
```
```suggestion
- If a RabbitMQ container does not correspond to a management API endpoint,
```

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+### Get the management endpoint URLs of application containers
+
+Next, we will write a function that fetches the URLs of our RabbitMQ containers
+.
```
Move the period to the previous line.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+			continue
+		}
+
+		u, err := url.Parse(fmt.Sprintf(
```
Use `net.JoinHostPort` instead of `Sprintf` here.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
````diff
@@ -0,0 +1,988 @@
+```go
+func cryptoRandomHex(len int) (string, error) {
+	randomBytes := make([]byte, len)
+	if _, err := rand.Reader.Read(randomBytes); err != nil {
````
You can just say `rand.Read(randomBytes)` here.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+
+`cryptoRandomHex` is a copy of a function that Teleport defines internally. It
+uses the `crypto/rand` package to generate random bytes, then converts them to a
+string in hexidecimal format, which we will use for the join token. You can use
```
```suggestion
string in hexadecimal format, which we will use for the join token. You can use
```

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+			Roles: types.SystemRoles{types.RoleApp},
+		})
+
+	if err := t.teleportClient.CreateToken(ctx, tok); err != nil {
```
This example might be simpler if you registered dynamic apps rather than started new app services and generated new join tokens.

ptgott replied:
I wanted to demonstrate the token method in this guide somehow so users could see an example, but it makes sense to advise using a dynamic approach for simplicity. I've moved the `Details` box about the dynamic approach higher in this section and changed the first paragraph after the code snippet to make the dynamic approach more prominent.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Deleted unnecessary Application Service record: " + p.GetName())
```
```suggestion
	fmt.Println("Deleted unnecessary Application Service record:" + p.GetName())
```

Here and below - Println already separates elements by space.

zmb3 commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,988 @@
+	fmt.Println("Starting the application")
+	app := newTokenDemoApp()
+
+	k := time.NewTicker(updateInterval)
```
```suggestion
	k := time.NewTicker(updateInterval)
	defer k.Stop()
```

ptgott committed:
Respond to zmb3 feedback

ptgott commented:
@alexfornuto This is ready for copy review

alexfornuto commented:
@ptgott do you wanna resolve the conflict with `introduction.mdx` before I review that page, in case the resolution changes the copy?

alexfornuto commented on docs/pages/api/automatically-register-agents.mdx:
````diff
@@ -0,0 +1,1023 @@
+$ for i in {1..3}; do docker run -d rabbitmq:3-management; done;
+```
+
+The terminal where you ran the application should show output similar to the
````
This is where my testing fails. I'm on macOS, and running the rabbit containers:

```
❯ docker container ps
CONTAINER ID   IMAGE                   COMMAND                  CREATED         STATUS         PORTS                                                                  NAMES
e20bbe8e9dea   rabbitmq:3-management   "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes   4369/tcp, 5671-5672/tcp, 15671-15672/tcp, 15691-15692/tcp, 25672/tcp   naughty_shtern
fbca4f93ba67   rabbitmq:3-management   "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes   4369/tcp, 5671-5672/tcp, 15671-15672/tcp, 15691-15692/tcp, 25672/tcp   affectionate_dubinsky
0f3a961dd928   rabbitmq:3-management   "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes   4369/tcp, 5671-5672/tcp, 15671-15672/tcp, 15691-15692/tcp, 25672/tcp   cranky_wright
```

But the go program is not responding:

```
❯ go run main.go
Starting the application
Connected to Teleport
Connected to the Docker daemon
```

And `tsh apps ls` doesn't list any new applications. I tried restarting the go client while the containers were still running, and then restarting the containers while the go client was running. I've also waited up to five minutes after creating the containers for the client app to do something.

Here's my completed main.go, to check for PEBKAC errors:

<details>
<summary>main.go</summary> 

```go
package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net"
	"net/url"
	"strings"
	"time"

	dtypes "github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/api/types/filters"
	"github.com/docker/docker/api/types/strslice"
	docker "github.com/docker/docker/client"
	teleport "github.com/gravitational/teleport/api/client"
	"github.com/gravitational/teleport/api/client/proto"
	"github.com/gravitational/teleport/api/types"
	"github.com/gravitational/trace"
	"google.golang.org/grpc"
)

const (
	// Assign proxyAddr to the host and port of your Teleport Proxy Service instance
	proxyAddr      string = "REDACTED:443"
	teleportImage  string = "public.ecr.aws/gravitational/teleport:12.1.1"
	initTimeout           = time.Duration(30) * time.Second
	updateInterval        = time.Duration(5) * time.Second
	tokenTTL              = time.Duration(5) * time.Minute
	networkName    string = "bridge"
	managementPort string = "15672"
	tokenLenBytes         = 16
	rabbitMQImage  string = "rabbitmq:3-management"
)

type tokenDemoApp struct {
	dockerClient   *docker.Client
	teleportClient *teleport.Client
}

func (t *tokenDemoApp) listRegisteredAppURLs(ctx context.Context) (map[string]types.AppServer, error) {
	m := make(map[string]types.AppServer)

	for {
		req := proto.ListResourcesRequest{
			ResourceType: "app_server",
			Limit:        10,
		}
		resp, err := t.teleportClient.ListResources(
			ctx,
			req,
		)

		if err != nil {
			return nil, trace.Wrap(err)
		}

		for _, r := range resp.Resources {
			if p, ok := r.(types.AppServer); ok {
				m[p.GetApp().GetURI()] = p
			}
		}

		// No more pages to request
		if resp.NextKey == "" {
			break
		}

		req.StartKey = resp.NextKey
	}

	return m, nil
}

func (t *tokenDemoApp) listAppContainerURLs(ctx context.Context, image string) (map[string]struct{}, error) {
	c, err := t.dockerClient.ContainerList(ctx, dtypes.ContainerListOptions{
		Filters: filters.NewArgs(filters.KeyValuePair{
			Key:   "ancestor",
			Value: image,
		}),
	})
	if err != nil {
		return nil, trace.Wrap(err)
	}

	l := make(map[string]struct{})

	for _, r := range c {
		b, ok := r.NetworkSettings.Networks[networkName]
		// Not connected to the chosen network, so skip it
		if !ok {
			continue
		}

		u, err := url.Parse("http://" + net.JoinHostPort(
			b.IPAddress,
			managementPort,
		))

		if err != nil {
			return nil, trace.Wrap(err)
		}

		l[u.String()] = struct{}{}
	}

	return l, nil
}

func cryptoRandomHex(len int) (string, error) {
	randomBytes := make([]byte, len)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", trace.Wrap(err)
	}
	return hex.EncodeToString(randomBytes), nil
}

func (t *tokenDemoApp) createAppToken(ctx context.Context) (string, error) {
	n, err := cryptoRandomHex(tokenLenBytes)
	if err != nil {
		return "", trace.Wrap(err)
	}

	tok, err := types.NewProvisionTokenFromSpec(
		n,
		time.Now().Add(tokenTTL),
		types.ProvisionTokenSpecV2{
			Roles: types.SystemRoles{types.RoleApp},
		})

	if err := t.teleportClient.CreateToken(ctx, tok); err != nil {
		return "", trace.Wrap(err)
	}
	return n, nil
}

func (t *tokenDemoApp) startApplicationServiceContainer(
	ctx context.Context,
	token string,
	u url.URL,
) error {

	name := strings.ReplaceAll(u.Hostname(), ".", "-")
	resp, err := t.dockerClient.ContainerCreate(
		ctx,
		&container.Config{
			Image: teleportImage,
			Entrypoint: strslice.StrSlice{
				"/usr/bin/dumb-init",
				"teleport",
				"start",
				"--roles=app",
				"--auth-server=" + proxyAddr,
				"--token=" + token,
				"--app-name=rabbitmq-" + name,
				"--app-uri=" + u.String(),
			},
		},
		nil,
		nil,
		nil,
		"",
	)
	if err != nil {
		return trace.Wrap(err)
	}

	err = t.dockerClient.ContainerStart(
		ctx,
		resp.ID,
		dtypes.ContainerStartOptions{},
	)
	if err != nil {
		return trace.Wrap(err)
	}

	return nil
}

func (t *tokenDemoApp) pruneAppServiceInstance(ctx context.Context, p types.AppServer) error {
	host := p.GetHostname()

	if err := t.teleportClient.DeleteApplicationServer(
		ctx,
		p.GetNamespace(),
		p.GetHostID(),
		p.GetName(),
	); err != nil {
		return trace.Wrap(err)
	}

	fmt.Println("Deleted unnecessary Application Service record:", p.GetName())

	// Don't check errors when removing the container, since it may already
	// have been removed.
	t.dockerClient.ContainerStop(ctx, host, container.StopOptions{})
	t.dockerClient.ContainerRemove(ctx, host, dtypes.ContainerRemoveOptions{})

	fmt.Println("Deleted unnecessary Application Service container:", host)
	return nil
}

func (t *tokenDemoApp) reconcileApps() error {
	ctx := context.Background()
	apps, err := t.listRegisteredAppURLs(ctx)
	if err != nil {
		return trace.Wrap(err)
	}

	urls, err := t.listAppContainerURLs(ctx, rabbitMQImage)
	if err != nil {
		return trace.Wrap(err)
	}

	for u, _ := range urls {
		if _, ok := apps[u]; ok {
			continue
		}
		tok, err := t.createAppToken(ctx)
		if err != nil {
			return trace.Wrap(err)
		}
		fmt.Println("Created a new application token for URL: " + u)

		r, err := url.Parse(u)
		if err != nil {
			return trace.Wrap(err)
		}

		err = t.startApplicationServiceContainer(ctx, tok, *r)
		if err != nil {
			return trace.Wrap(err)
		}
		fmt.Println("Started an Application Service container to proxy URL: " + u)
	}

	for a, p := range apps {
		_, ok := urls[a]
		if ok {
			continue
		}

		if err := t.pruneAppServiceInstance(ctx, p); err != nil {
			return trace.Wrap(err)
		}
	}
	return nil
}

func newTokenDemoApp() *tokenDemoApp {
	ctx := context.Background()
	ctx, cancel := context.WithTimeout(ctx, initTimeout)
	defer cancel()
	creds := teleport.LoadIdentityFile("auth.pem")

	t, err := teleport.New(ctx, teleport.Config{
		Addrs:       []string{proxyAddr},
		Credentials: []teleport.Credentials{creds},
		DialOpts: []grpc.DialOption{
			grpc.WithReturnConnectionError(),
		},
	})
	if err != nil {
		panic(err)
	}
	fmt.Println("Connected to Teleport")

	d, err := docker.NewClientWithOpts(
		docker.WithAPIVersionNegotiation(),
	)
	if err != nil {
		panic(err)
	}
	fmt.Println("Connected to the Docker daemon")

	return &tokenDemoApp{
		teleportClient: t,
		dockerClient:   d,
	}

}

func main() {
	fmt.Println("Starting the application")
	app := newTokenDemoApp()

	k := time.NewTicker(updateInterval)
	defer k.Stop()
	for {
		<-k.C
		if err := app.reconcileApps(); err != nil {
			panic(err)
		}
	}
}
```

</details>

ptgott replied:
I'm going to push a commit that includes the whole program in the `examples` directory so there's no risk of accidentally pasting the wrong thing (I don't think that that's the problem, but just in case...). 

It works on my machine™, so there must be some difference between our environments that's coming into play. 

One thing you could do is add some `Println` calls to `*tokenDemoApp.reconcileApps` to see if the app is listing containers/registered apps as expected:

```diff
func (t *tokenDemoApp) reconcileApps() error {
	ctx := context.Background()
	apps, err := t.listRegisteredAppURLs(ctx)
	if err != nil {
		return trace.Wrap(err)
	}

+     fmt.Println("registered apps", apps)

	urls, err := t.listAppContainerURLs(ctx, rabbitMQImage)
	if err != nil {
		return trace.Wrap(err)
	}

+      fmt.Println("app container URLs:", urls)

	for u, _ := range urls {
		if _, ok := apps[u]; ok {
			continue
		}
		tok, err := t.createAppToken(ctx)
		if err != nil {
			return trace.Wrap(err)
		}
		fmt.Println("Created a new application token for URL: " + u)

		r, err := url.Parse(u)
		if err != nil {
			return trace.Wrap(err)
		}

		err = t.startApplicationServiceContainer(ctx, tok, *r)
		if err != nil {
			return trace.Wrap(err)
		}
		fmt.Println("Started an Application Service container to proxy URL: " + u)
	}

	for a, p := range apps {
		_, ok := urls[a]
		if ok {
			continue
		}

		if err := t.pruneAppServiceInstance(ctx, p); err != nil {
			return trace.Wrap(err)
		}
	}
	return nil
}
```

Are the containers connected to the default bridge network? 

```
docker network inspect bridge | jq '.[0].Containers'
```

If not, the app also won't be able to find them, since it only keeps track of container URLs on the default bridge network.

In any case, thanks for testing this out!

alexfornuto reviewed (commented):
Some additional copy edits.

alexfornuto commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,1023 @@
+
+The client imports the following third-party code:
+
+|Package|Description|
+|---|---|
+|`github.com/docker/docker/api/types`|Types used in the Docker daemon API. Aliased as `dtypes` here. |
+|`github.com/docker/docker/api/types/container`|Container-specific types used in the Docker daemon API.|
+|`github.com/docker/docker/api/types/filters`|Types used for filtering containers in the Docker daemon API.|
+|`github.com/docker/docker/api/types/strslice`|A utility package for working with slices of strings in Docker's API client library. (In Go, slices are similar to arrays, but with variable lengths and capacities. Arrays have a fixed size.)|
+|`github.com/docker/docker/client`|The Docker API client library, aliased as `docker` here.|
+|`github.com/gravitational/teleport/api/client`|A library for authenticating to the Auth Service's gRPC API and making requests, aliased as `teleport`.|
+|`github.com/gravitational/api/client/proto`|Teleport's protocol buffer API specification.|
+|`github.com/gravitational/teleport/api/types`|Types used in the Auth Service API, e.g., Application Service records.|
+|`github.com/gravitational/trace`|Presenting errors with more useful detail than the standard library provides.|
+|`google.golang.org/grpc`|The gRPC client and server library.|
```
When rendered:

![image](https://user-images.githubusercontent.com/2349184/225660879-396c5353-0846-4f74-a0a0-0767d6f0f34e.png)

ptgott replied:
Adding the `nobr` tags around the `github.com/docker/docker/api/types/container` path, the longest one in this table, worked for me! The second column is still readable, I think.

ptgott committed:
Use the examples directory for the code

This way, users can have a compilable example before they start working
through the guide.

ptgott commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,1023 @@
+
+### Create a user and role for the client application
+
+When writing a Teleport API client application, you must define a Teleport role
+that gives the application adequate permissions for the API operations it needs
+to perform.
```
I've removed the paragraph and tweaked the paragraph after it into the first paragraph of this H3 section.

ptgott commented on docs/pages/api/automatically-register-agents.mdx:
These are from my `gofmt`- formatted `main.go` file. I relented from changing these since they began to render oddly on the preview site, but I can put in the effort if it makes sense to you.

ptgott commented on docs/pages/api/automatically-register-agents.mdx:
````diff
@@ -0,0 +1,1023 @@
+method:
+
+```go
+func (c *Client) CreateApp(ctx context.Context, app types.Application) error
````
The current app creates a token and launches an App Service instance with it. If you use `CreateApp`, you don't need to create a token or launch a new App Service instance, but you do need to have an App Service instance running already.

ptgott commented on docs/pages/api/automatically-register-agents.mdx:
````diff
@@ -0,0 +1,1023 @@
+```
+
+For this to work, you must have an instance of the Application Service already
+running. Your client application must also have the following permissions:
````
I've changed this to "Your API client application's Teleport user must also have the following permissions:"

ptgott commented on docs/pages/api/automatically-register-agents.mdx:
````diff
@@ -0,0 +1,1023 @@
+create containers based on it:
+
+```code
+$ docker image pull (=teleport.latest_oss_docker_image=)
````
It didn't when I ran the demo app. I think the `docker` CLI bundles some actions together that the API client library separates. `docker run` pulls a container, creates it, and executes it, whereas these are separate Docker API client function calls. I thought it would make sense to pull the container manually, since it's a quick step that you only need to do once.

ptgott commented on docs/pages/api/introduction.mdx:
```diff
@@ -4,24 +4,43 @@ description: Introduction to the Teleport gRPC API.
+ - Integrate with external tools, e.g., to write an [Access Request
+   plugin](../access-controls/access-request-plugins/index.mdx). Teleport
+   maintains Access Request plugins for tools like Slack, Jira, and Mattermost.
+ - Writing a program/bot to manage Access Requests automatically, based on your
```
👍 
I've also deleted this particular line (from the original page), since we don't recommend approving/denying Access Requests automatically (and recommend displaying links to the Teleport Web UI to do this).

ptgott committed:
Respond to alexfornuto feedback

ptgott commented:
@alexfornuto Thanks for the review! I've responded to your feedback and also pushed a commit to incorporate some feedback from Zac [in this PR](https://github.com/gravitational/teleport/pull/23204) that also applies to this one.

alexfornuto commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,1023 @@
+
+<Notice type="danger">
+
+The program we will build in this guide is intended as a learning tool. **Do not
```
This might be over-jealousness on my part. I'm trying to follow the advice from [this talk](https://writing.rocks/to-be-or-not-to-be/), and the broader advice in @MarciaRieferJohnston's book. In this example, "will" is a "weak be verb" in that it's shorthand for "will be".  

In general, I'm always trying to reduce the amount of words needed to say something. I think about how I would say it if I were speaking aloud. It's probably that last point that most compels me to make suggestions like this.

It's your call buddy, do as you will.

MarciaRieferJohnston replied:
@alexfornuto Hi, Alex. This is a thoughtful suggestion, as always. Nice to be remembered. :) 

@ptgott Hi, Paul. Good question. Future tense ("will" + verb) isn't wrong. And you're right about the subject ("we") still performing the action with or without the "will." 

FWIW, here's one person's way of describing the reason for preferring present tense to future tense, which is generally considered a best practice (but not a have-to-do): 

> "In technical content, the future tense does not have the same immediacy as the present tense. Compare 'When you connect the power cable, the device will boot up automatically,' to 'When you connect the power cable, the device boots up automatically.' The ... present tense conveys immediacy and demonstrates confidence."

In sentences like yours, which sound fine either way (with or without the "will"), it's a judgment call. Tighter is generally better and stronger, so I'd leave it out. In cases where something will happen later and the relative time difference makes it weird to use present tense, though, then sure, use "will." As Alex says, it's up to you.

ptgott replied:
Thanks for taking the time to contribute your thoughts!

alexfornuto commented on docs/pages/api/automatically-register-agents.mdx:
```diff
@@ -0,0 +1,979 @@
+|Resource|Method|Within a Role|
+|---|---|---|
+|Database|`*client.Client.CreateDatabase`|`db`|
+|Kubernetes cluster|`*client.Client.CreateKubernetesCluster`|`kube_cluster`|
+|Windows Desktop|`*client.Client.CreateWindowsDesktop`|`windows_desktop`|
```
```suggestion
|Kubernetes cluster|<nobr>`*client.Client.CreateKubernetesCluster`</nobr>|`kube_cluster`|
|Windows Desktop|`*client.Client.CreateWindowsDesktop`|<nobr>`windows_desktop`</nobr>|
```

alexfornuto reviewed (approved):
Just a couple more suggestions.

ptgott committed:
Respond to PR feedback

ptgott committed:
Run make fix-license
